File: exp9/scan.py
def max_num(head, sequence):
    for i in sequence:
        if i > head:
            return i
def min_num(head, sequence):
    for i in range(len(sequence) - 1, -1, -1):
        if sequence[i] < head:
            return sequence[i]
def SCAN(N, head, sequence):
    old_head = head
    seek_sequence = []
    seek_operations = 0
    seek_sequence.append(head)
    near_num = max_num(head, sequence)
    for i in range(len(sequence)):
        if near_num > head:
            difference = near_num - head
            seek_operations += difference
            head = near_num
            seek_sequence.append(head)
            near_num = max_num(head, sequence)
        elif near_num < head:
            difference = head - near_num
            seek_operations += difference
            head = near_num
            seek_sequence.append(head)
            near_num = min_num(head, sequence)
            if head == min(sequence):
                break
        if head == max(sequence):
            near_num = min_num(old_head, sequence)
            difference = (N - 1 - head) + (N - 1 - near_num)
            seek_operations += difference
            head = near_num
            seek_sequence.append(head)
            near_num = min_num(head, sequence)
    print("Seek Sequence : 	", end=" ")
    for i in seek_sequence:
        if i == min(seek_sequence):
            print(i)
        elif i == max(sequence):
            print(i, " ==> ", N - 1, " ==> ", end=" ")
        else:
            print(i, " ==> ", end=" ")
    return seek_operations
if __name__ == "__main__":
    Number_disk = int(input("Enter the number of disks:	"))
    if Number_disk > 0:
        head = int(input("Enter initial header position : "))
        while not head in range(Number_disk + 1):
            head = int(input("Please enter valid initial head position:"))
        sequence = []
        sequence = list(map(int, input("Enter the sequence :").split()))
        sequence.sort()
        if min(sequence) < 0 or max(sequence) > Number_disk:
            print("Sequence out of range")
            exit(0)

        seek_operations = SCAN(Number_disk, head, sequence)
        print("Total number of seek operations : ", seek_operations)

File: exp9/scan.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>

int absoluteVal(int x){
    if (x>0) return x;
    else return x*-1;
}
void main(){
    int queue[30],n,head,i,j,k,seek=0,range,diff,temp,q1[20],q2[20],temp1=0,temp2=0;
    printf("SCAN algo: \n");
    printf("Enter max range of disk: ");
    scanf("%d",&range);
    printf("\nEnter no. of queue request: ");
    scanf("%d",&n);
    printf("\nEnter initial head position: ");
    scanf("%d",&head);
    printf("Enter queue sequence: ");
    for ( i = 1; i <=n; i++)
    {
        scanf("%d",&temp);
        if(temp>head){
            q1[temp1] = temp;
            temp1++;
        }
        else{
            q2[temp2] = temp;
            temp2++;
        }
    }
    // sorting in ascending order
    for(i=0;i<temp1-1;i++){
        for(j=i+1;j<temp1;j++){
            if(q1[i]>q1[j]){
                temp = q1[i];
                q1[i] = q1[j];
                q1[j] = temp;
            }
        }
    }
    // sorting in descending order
    for(i=0;i<temp2-1;i++){
        for(j=i+1;j<temp2;j++){
            if(q2[i]<q2[j]){
                temp = q2[i];
                q2[i] = q2[j];
                q2[i]=temp;
            }
        }
    }
    // copy q1 contents to queue
    for( i = 1,j=0; j < temp1; i++,j++)
    {
        queue[i]=q1[j];
    }
    queue[i]=range;
    // copy q2 contents to queue
    for( i = temp1+2,j=0; j < temp2; i++,j++)
    {
        queue[i]=q2[j];
    }
    queue[i]=0;
    queue[0]=head;
    for (j = 0; j <=n; j++)
    {
        diff = absoluteVal(queue[j+1]-queue[j]);
        seek+=diff;
        printf("\n Disk moves from %d => %d with seek time: %d \n",queue[j],queue[j+1],diff);
    }
    
}